File: typescript/src/core/xbrl/instance.ts
/**
 * XBRL instance of the high-level API.
 *
 * `XBRLInstance` composes two data sources for one filing:
 *
 * - fact-level queries served from the company-facts API (`queries`), and
 * - filing-scoped statement structure parsed from the filing's rendered
 *   reports (`rendered-reports`), including dimensional detail such as
 *   segment breakdowns.
 */

import {
  XbrlFact,
  XbrlCompanyFacts,
  XbrlQueryOptions,
  FinancialStatement,
  BalanceSheet,
  IncomeStatement,
  CashFlowStatement,
  BalanceSheetItem,
  IncomeStatementItem,
  CashFlowItem,
} from "../../types/xbrl";
import {
  buildFactRecords,
  factsHistory,
  parseFilterExpression,
  FactHistoryRow,
  FactQuery,
} from "./queries";
import { AsReportedStatements } from "./as-reported";
import { RenderedReportReader, ReportDescriptor } from "./rendered-reports";
import { StatementLineItem } from "./report-html-parser";
import { STATEMENT_CONCEPTS, normalizeStatementType } from "./statements";

/** The pieces of a filing (fluent or object API) the instance needs. */
export interface XbrlFilingLike {
  cik: string;
  accessionNumber?: string;
  formType?: string;
  filingDate?: string | Date;
  api?: XbrlApiLike;
}

/** The pieces of the API client the instance needs. */
export interface XbrlApiLike {
  xbrl: { getCompanyFacts(cik: string | number): Promise<Record<string, any>> };
  httpClient?: { getRaw(url: string): Promise<string> };
  getFiling?(
    cik: string | number,
    accessionNumber: string,
  ): Promise<Record<string, any>>;
}

export class XBRLInstance {
  private _api: XbrlApiLike;
  private filing: XbrlFilingLike;
  public readonly cik: string;

  // Cache for XBRL data
  private _facts: XbrlCompanyFacts | null = null;
  private _usGaapFacts: Record<string, any> | null = null;
  private _deiFacts: Record<string, any> | null = null;
  private _reportsReader: RenderedReportReader | null = null;
  private _asReported: AsReportedStatements | null = null;

  constructor(filing: XbrlFilingLike, api?: XbrlApiLike) {
    const resolvedApi = api || filing.api;
    if (!resolvedApi) {
      throw new Error("XBRLInstance requires an API client");
    }
    this._api = resolvedApi;
    this.filing = filing;
    this.cik = filing.cik;
  }

  private get archiveBase(): string {
    const accession = (this.filing.accessionNumber || "").replace(/-/g, "");
    return `https://www.sec.gov/Archives/edgar/data/${parseInt(this.cik, 10)}/${accession}`;
  }

  /** Reader for the filing's rendered reports. */
  get reports(): RenderedReportReader {
    if (!this._reportsReader) {
      const http = this._api.httpClient;
      if (!http) {
        throw new Error("Rendered-report access requires an HTTP client");
      }
      this._reportsReader = new RenderedReportReader(this.archiveBase, http);
    }
    return this._reportsReader;
  }

  /**
   * Statements assembled from the filing's own XBRL fileset
   * (instance document plus presentation and label linkbases).
   */
  async asReported(): Promise<AsReportedStatements> {
    if (!this._asReported) {
      const fileNames: string[] = [];
      if (this._api.getFiling && this.filing.accessionNumber) {
        try {
          const details = await this._api.getFiling(
            this.cik,
            this.filing.accessionNumber,
          );
          for (const item of details?.directory?.item || []) {
            if (item?.name) {
              fileNames.push(String(item.name));
            }
          }
        } catch {
          // fall through with an empty listing
        }
      }
      const http = this._api.httpClient;
      if (!http) {
        throw new Error("As-reported access requires an HTTP client");
      }
      this._asReported = new AsReportedStatements(
        this.archiveBase,
        http,
        fileNames,
      );
    }
    return this._asReported;
  }

  /** The parsed XBRL instance document, or null when absent. */
  async instanceDocument() {
    return (await this.asReported()).getDocument();
  }

  /**
   * Get all XBRL facts for the company
   */
  async getFacts(): Promise<XbrlCompanyFacts> {
    if (!this._facts) {
      this._facts = (await this._api.xbrl.getCompanyFacts(
        this.cik,
      )) as XbrlCompanyFacts;
    }
    return this._facts;
  }

  /**
   * Get US-GAAP facts
   */
  async getUsGaap(): Promise<Record<string, any>> {
    if (!this._usGaapFacts) {
      const facts = await this.getFacts();
      this._usGaapFacts = facts?.facts?.["us-gaap"] || {};
    }
    return this._usGaapFacts;
  }

  /**
   * Get DEI (Document Entity Information) facts
   */
  async getDei(): Promise<Record<string, any>> {
    if (!this._deiFacts) {
      const facts = await this.getFacts();
      this._deiFacts = facts?.facts?.dei || {};
    }
    return this._deiFacts;
  }

  /**
   * Query XBRL facts with filtering.
   *
   * Accepts an options object (`{ concept: 'Assets', unit: 'USD' }`) or a
   * filter expression string (`"concept=Assets&unit=USD"`; the empty
   * string selects everything). Returns a FactQuery (an array with a
   * chainable `byConcept()` helper).
   */
  async query(options: XbrlQueryOptions | string = {}): Promise<FactQuery> {
    if (typeof options === "string") {
      options = options === "" ? {} : parseFilterExpression(options);
    }
    const { concept, taxonomy = "us-gaap", unit, period } = options;

    const results: XbrlFact[] = [];

    // Get facts for the specified taxonomy
    let taxonomyFacts: Record<string, any>;
    if (taxonomy === "us-gaap") {
      taxonomyFacts = await this.getUsGaap();
    } else if (taxonomy === "dei") {
      taxonomyFacts = await this.getDei();
    } else {
      const facts = await this.getFacts();
      taxonomyFacts = facts?.facts?.[taxonomy] || {};
    }

    if (concept) {
      if (concept in taxonomyFacts) {
        results.push(
          ...buildFactRecords(
            concept,
            taxonomyFacts[concept],
            taxonomy,
            unit,
            period,
          ),
        );
      }
    } else {
      for (const [conceptName, conceptData] of Object.entries(taxonomyFacts)) {
        results.push(
          ...buildFactRecords(conceptName, conceptData, taxonomy, unit, period),
        );
      }
    }

    return FactQuery.fromRecords(results);
  }

  /** History of one concept, sorted by period end ascending. */
  async factsHistory(
    concept: string,
    taxonomy: string = "us-gaap",
  ): Promise<FactHistoryRow[]> {
    const facts = await this.getFacts();
    return factsHistory(facts as Record<string, any>, concept, taxonomy);
  }

  // ------------------------------------------------------------------
  // Filing-scoped statements (rendered reports)
  // ------------------------------------------------------------------

  /**
   * List all rendered reports in the filing, with `definition`,
   * `shortName`, `role`, and `rFile` keys.
   */
  async getAllStatements(): Promise<ReportDescriptor[]> {
    const rendered = await this.reports.listReports();
    if (rendered.length > 0) {
      return rendered;
    }
    // No FilingSummary (the renderer never ran for this filing):
    // fall back to the roles defined by the presentation linkbase.
    const asReported = await this.asReported();
    if (asReported.isAvailable) {
      const roles = await asReported.listRoles();
      return roles.map((role) => ({
        definition: role.split("/").pop() || role,
        shortName: role.split("/").pop() || role,
        role,
        rFile: "",
        reportType: "",
      }));
    }
    return rendered;
  }

  /**
   * Parse one rendered report into line items.
   *
   * @param role The report's role URI (from `getAllStatements()`), its
   *   R-file name, or its short name.
   */
  async getStatement(
    role: string,
    source: "auto" | "instance" | "rendered" = "auto",
  ): Promise<StatementLineItem[]> {
    if (source !== "rendered") {
      const asReported = await this.asReported();
      if (asReported.isAvailable) {
        let resolved = role;
        if (role.toLowerCase().endsWith(".htm")) {
          const report = await this.reports.findReport(role);
          if (report) {
            resolved = report.role;
          }
        }
        const items = await asReported.getStatement(resolved);
        if (items.length > 0 || source === "instance") {
          return items;
        }
      }
      if (source === "instance") {
        return [];
      }
    }
    return this.reports.readStatement(role);
  }

  /**
   * Find and extract a specific financial statement.
   *
   * Accepts "balance_sheet" / "income_statement" / "cash_flow" and
   * common CamelCase spellings ("BalanceSheet").
   */
  async findStatement(
    statementType: string,
    period?: string,
  ): Promise<FinancialStatement | null> {
    const normalized = normalizeStatementType(statementType);
    const concepts = STATEMENT_CONCEPTS[normalized];
    if (!concepts) {
      return null;
    }
    return this.extractStatementFromFacts(normalized, concepts, period);
  }

  private async extractStatementFromFacts(
    statementType: string,
    concepts: string[],
    period?: string,
  ): Promise<FinancialStatement> {
    const statementData: Record<string, XbrlFact> = {};

    for (const concept of concepts) {
      let facts = await this.query({ concept, unit: "USD", period });
      if (facts.length === 0) {
        facts = await this.query({ concept, period });
      }
      if (facts.length > 0) {
        const latestFact = facts.reduce((prev, current) =>
          (current.filed || "") > (prev.filed || "") ? current : prev,
        );
        statementData[concept] = latestFact;
      }
    }

    return {
      statement_type: statementType as FinancialStatement["statement_type"],
      period,
      data: statementData,
    };
  }

  /**
   * Get structured balance sheet
   */
  async getBalanceSheet(period?: string): Promise<BalanceSheet> {
    const statement = await this.findStatement("balance_sheet", period);
    const data = statement?.data || {};

    const toBalanceSheetItem = (fact: XbrlFact): BalanceSheetItem => ({
      label: fact.concept,
      value:
        typeof fact.value === "number"
          ? fact.value
          : parseFloat(fact.value as string),
      units: fact.unit,
      period: fact.period,
      filed: new Date(fact.filed || ""),
    });

    const assets = {
      current_assets: data.AssetsCurrent
        ? [toBalanceSheetItem(data.AssetsCurrent)]
        : [],
      non_current_assets: data.AssetsNoncurrent
        ? [toBalanceSheetItem(data.AssetsNoncurrent)]
        : [],
      total_assets: data.Assets ? toBalanceSheetItem(data.Assets) : undefined,
    };

    const liabilities = {
      current_liabilities: data.LiabilitiesCurrent
        ? [toBalanceSheetItem(data.LiabilitiesCurrent)]
        : [],
      non_current_liabilities: data.LiabilitiesNoncurrent
        ? [toBalanceSheetItem(data.LiabilitiesNoncurrent)]
        : [],
      total_liabilities: data.Liabilities
        ? toBalanceSheetItem(data.Liabilities)
        : undefined,
    };

    const equity = {
      total_equity: data.StockholdersEquity
        ? toBalanceSheetItem(data.StockholdersEquity)
        : undefined,
      retained_earnings: data.RetainedEarningsAccumulatedDeficit
        ? toBalanceSheetItem(data.RetainedEarningsAccumulatedDeficit)
        : undefined,
    };

    return { assets, liabilities, equity };
  }

  /**
   * Get structured income statement
   */
  async getIncomeStatement(period?: string): Promise<IncomeStatement> {
    const statement = await this.findStatement("income_statement", period);
    const data = statement?.data || {};

    const toIncomeStatementItem = (fact: XbrlFact): IncomeStatementItem => ({
      label: fact.concept,
      value:
        typeof fact.value === "number"
          ? fact.value
          : parseFloat(fact.value as string),
      units: fact.unit,
      period: fact.period,
      filed: new Date(fact.filed || ""),
    });

    return {
      revenue:
        data.Revenues ||
        data.RevenueFromContractWithCustomerExcludingAssessedTax
          ? toIncomeStatementItem(
              data.Revenues ||
                data.RevenueFromContractWithCustomerExcludingAssessedTax,
            )
          : undefined,
      gross_profit: data.GrossProfit
        ? toIncomeStatementItem(data.GrossProfit)
        : undefined,
      operating_income: data.OperatingIncomeLoss
        ? toIncomeStatementItem(data.OperatingIncomeLoss)
        : undefined,
      net_income: data.NetIncomeLoss
        ? toIncomeStatementItem(data.NetIncomeLoss)
        : undefined,
      earnings_per_share: data.EarningsPerShareBasic
        ? toIncomeStatementItem(data.EarningsPerShareBasic)
        : undefined,
      operating_expenses: [],
    };
  }

  /**
   * Get structured cash flow statement
   */
  async getCashFlowStatement(period?: string): Promise<CashFlowStatement> {
    const statement = await this.findStatement("cash_flow", period);
    const data = statement?.data || {};

    const toCashFlowItem = (fact: XbrlFact): CashFlowItem => ({
      label: fact.concept,
      value:
        typeof fact.value === "number"
          ? fact.value
          : parseFloat(fact.value as string),
      units: fact.unit,
      period: fact.period,
      filed: new Date(fact.filed || ""),
    });

    return {
      operating_activities: data.NetCashProvidedByUsedInOperatingActivities
        ? [toCashFlowItem(data.NetCashProvidedByUsedInOperatingActivities)]
        : [],
      investing_activities: data.NetCashProvidedByUsedInInvestingActivities
        ? [toCashFlowItem(data.NetCashProvidedByUsedInInvestingActivities)]
        : [],
      financing_activities: data.NetCashProvidedByUsedInFinancingActivities
        ? [toCashFlowItem(data.NetCashProvidedByUsedInFinancingActivities)]
        : [],
      net_cash_flow:
        data.CashCashEquivalentsRestrictedCashAndRestrictedCashEquivalents
          ? toCashFlowItem(
              data.CashCashEquivalentsRestrictedCashAndRestrictedCashEquivalents,
            )
          : undefined,
    };
  }

  /**
   * Get a single value for a specific concept
   */
  async getConceptValue(
    concept: string,
    taxonomy: string = "us-gaap",
    unit: string = "USD",
    period?: string,
  ): Promise<number | null> {
    const facts = await this.query({ concept, taxonomy, unit, period });
    if (facts.length > 0) {
      const latestFact = facts.reduce((prev, current) =>
        (current.filed || "") > (prev.filed || "") ? current : prev,
      );
      return typeof latestFact.value === "number"
        ? latestFact.value
        : parseFloat(latestFact.value as string);
    }
    return null;
  }

  /**
   * Get all facts for a specific concept
   */
  async getFactsByConcept(
    concept: string,
    taxonomy: string = "us-gaap",
  ): Promise<XbrlFact[]> {
    const facts = await this.getFacts();
    const taxonomyFacts = facts?.facts?.[taxonomy];

    if (!taxonomyFacts || !taxonomyFacts[concept]) {
      return [];
    }

    const conceptData = taxonomyFacts[concept];
    const results: XbrlFact[] = [];

    // Dated periods (end/instant), unlike query()'s fiscal-frame periods
    for (const [unit, values] of Object.entries(conceptData.units || {})) {
      if (Array.isArray(values)) {
        for (const value of values) {
          results.push({
            concept,
            taxonomy,
            value: value.val,
            unit,
            period: value.end || value.instant || "",
            fiscal_year: value.fy,
            fiscal_period: value.fp,
            start_date: value.start,
            end_date: value.end,
            filed: value.filed,
            accession_number: value.accn,
            form: value.form,
          });
        }
      }
    }

    return results;
  }

  /**
   * List all available concepts in a taxonomy
   */
  async listConcepts(taxonomy: string = "us-gaap"): Promise<string[]> {
    let taxonomyFacts: Record<string, any>;

    if (taxonomy === "us-gaap") {
      taxonomyFacts = await this.getUsGaap();
    } else if (taxonomy === "dei") {
      taxonomyFacts = await this.getDei();
    } else {
      const facts = await this.getFacts();
      taxonomyFacts = facts?.facts?.[taxonomy] || {};
    }

    return Object.keys(taxonomyFacts);
  }

  /**
   * Convert XBRL data to plain object
   */
  async toObject(options: XbrlQueryOptions = {}): Promise<any> {
    const facts = await this.query(options);
    return {
      metadata: {
        cik: this.cik,
        filing_date: this.filing.filingDate,
        form_type: this.filing.formType,
      },
      facts: facts,
    };
  }

  toString(): string {
    return `XBRL instance for ${this.filing.formType} filing (CIK: ${this.cik})`;
  }
}

File: python/src/sec_edgar_toolkit/core/xbrl/instance.py
"""
XBRL instance of the high-level API.

``XBRLInstance`` composes two data sources for one filing:

- fact-level queries served from the company-facts API
  (:mod:`.queries`), and
- filing-scoped statement structure parsed from the filing's rendered
  reports (:mod:`.rendered_reports`).
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any, Dict, List, Optional

from ...client import SecEdgarApi
from .queries import FactQuery, FactsData, build_fact_records, parse_filter_expression
from .rendered_reports import RenderedReportReader
from .statements import STATEMENT_CONCEPTS, normalize_statement_type

if TYPE_CHECKING:
    from ..filing import Filing

logger = logging.getLogger(__name__)


class XBRLInstance:
    """
    XBRL data for one filing.

    Attributes:
        filing: Associated Filing object
        cik: Company's Central Index Key
        facts: Raw company-facts payload (with ``facts_history()`` helper)
    """

    def __init__(
        self,
        filing: "Filing",
        api: Optional[SecEdgarApi] = None,
    ) -> None:
        if api is None:
            from ..global_functions import _get_api

            api = _get_api()

        self._api = api
        self.filing = filing
        self.cik = filing.cik

        self._facts: Optional[FactsData] = None
        self._us_gaap_facts: Optional[Dict[str, Any]] = None
        self._dei_facts: Optional[Dict[str, Any]] = None
        self._reports_reader: Optional[RenderedReportReader] = None

    @property
    def facts(self) -> FactsData:
        """All XBRL facts for the company."""
        if self._facts is None:
            self._facts = FactsData(self._api.get_company_facts(self.cik))
        return self._facts

    @property
    def us_gaap(self) -> Dict[str, Any]:
        """Get US-GAAP facts."""
        if self._us_gaap_facts is None:
            self._us_gaap_facts = self.facts.get("facts", {}).get("us-gaap", {})
        return self._us_gaap_facts

    @property
    def dei(self) -> Dict[str, Any]:
        """Get DEI (Document Entity Information) facts."""
        if self._dei_facts is None:
            self._dei_facts = self.facts.get("facts", {}).get("dei", {})
        return self._dei_facts

    @property
    def reports(self) -> RenderedReportReader:
        """Reader for the filing's rendered reports."""
        if self._reports_reader is None:
            self._reports_reader = RenderedReportReader(
                self.filing._archive_base, self._api.http_client
            )
        return self._reports_reader

    # ------------------------------------------------------------------
    # Fact queries
    # ------------------------------------------------------------------

    def query(
        self,
        concept: Optional[str] = None,
        taxonomy: str = "us-gaap",
        unit: Optional[str] = None,
        period: Optional[str] = None,
        **kwargs: Any,
    ) -> FactQuery:
        """
        Query XBRL facts with filtering.

        The first argument accepts either a concept name ("Assets") or a
        filter expression ("concept=Assets&unit=USD"; the empty string
        selects everything).

        Returns:
            FactQuery (a list of fact records with ``to_dataframe()`` and
            ``by_concept()`` helpers)
        """
        # Filter-expression form: "concept=Assets" / ""
        if isinstance(concept, str):
            if concept == "":
                concept = None
            elif "=" in concept:
                concept, taxonomy, unit, period = parse_filter_expression(
                    concept, taxonomy, unit, period
                )

        if taxonomy == "us-gaap":
            taxonomy_facts = self.us_gaap
        elif taxonomy == "dei":
            taxonomy_facts = self.dei
        else:
            taxonomy_facts = self.facts.get("facts", {}).get(taxonomy, {})

        results: List[Dict[str, Any]] = []
        if concept:
            if concept in taxonomy_facts:
                results.extend(
                    build_fact_records(
                        concept, taxonomy_facts[concept], taxonomy, unit, period
                    )
                )
        else:
            for concept_name, concept_data in taxonomy_facts.items():
                results.extend(
                    build_fact_records(
                        concept_name, concept_data, taxonomy, unit, period
                    )
                )

        return FactQuery(results)

    # ------------------------------------------------------------------
    # Filing-scoped statements (rendered reports)
    # ------------------------------------------------------------------

    def get_all_statements(self) -> List[Dict[str, Any]]:
        """
        List all rendered reports in the filing.

        Returns:
            List of dicts with ``definition``, ``short_name``, ``role``,
            and ``r_file`` keys.
        """
        return self.reports.list_reports()

    def get_statement(self, role: str) -> List[Dict[str, Any]]:
        """
        Parse one rendered report into line items.

        Args:
            role: The report's role URI (from ``get_all_statements()``),
                its R-file name, or its short name.

        Returns:
            List of line-item dicts with ``concept``, ``label``,
            ``values`` (period -> number), ``units`` (period -> unit),
            and ``has_values`` keys.
        """
        return self.reports.read_statement(role)

    def find_statement(
        self,
        statement_type: str = "balance_sheet",
        period: Optional[str] = None,
    ) -> Optional[Dict[str, Any]]:
        """
        Find and extract a specific financial statement.

        Args:
            statement_type: "balance_sheet", "income_statement", or
                "cash_flow" (common CamelCase spellings are accepted too)
            period: Period filter (e.g., "2023", "CY2023Q4")

        Returns:
            Dictionary containing the financial statement data
        """
        statement_type = normalize_statement_type(statement_type)
        concepts = STATEMENT_CONCEPTS.get(statement_type)
        if concepts is None:
            logger.warning(f"Unknown statement type: {statement_type}")
            return None

        statement_data = {}
        for concept in concepts:
            facts = self.query(concept=concept, unit="USD", period=period)
            if not facts:
                facts = self.query(concept=concept, period=period)
            if facts:
                latest_fact = max(facts, key=lambda x: x.get("filed", "") or "")
                statement_data[concept] = latest_fact

        return {
            "statement_type": statement_type,
            "period": period,
            "data": statement_data,
        }

    # ------------------------------------------------------------------
    # Conversion helpers
    # ------------------------------------------------------------------

    def to_dataframe(
        self,
        concept: Optional[str] = None,
        taxonomy: str = "us-gaap",
        unit: Optional[str] = None,
    ):
        """Convert XBRL facts to a pandas DataFrame."""
        return self.query(concept=concept, taxonomy=taxonomy, unit=unit).to_dataframe()

    def to_dict(
        self,
        concept: Optional[str] = None,
        taxonomy: str = "us-gaap",
        unit: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Convert XBRL facts to dictionary format."""
        facts = self.query(concept=concept, taxonomy=taxonomy, unit=unit)
        return {
            "metadata": {
                "cik": self.cik,
                "filing_date": str(self.filing.filing_date),
                "form_type": self.filing.form_type,
            },
            "facts": list(facts),
        }

    def get_concept_value(
        self,
        concept: str,
        taxonomy: str = "us-gaap",
        unit: str = "USD",
        period: Optional[str] = None,
    ) -> Optional[float]:
        """Get the most recent value for a specific concept."""
        facts = self.query(concept=concept, taxonomy=taxonomy, unit=unit, period=period)
        if facts:
            latest_fact = max(facts, key=lambda x: x.get("filed", "") or "")
            return latest_fact.get("value")
        return None

    def list_concepts(self, taxonomy: str = "us-gaap") -> List[str]:
        """List all available concepts in a taxonomy."""
        if taxonomy == "us-gaap":
            return list(self.us_gaap.keys())
        elif taxonomy == "dei":
            return list(self.dei.keys())
        else:
            facts = self.facts.get("facts", {})
            return list(facts.get(taxonomy, {}).keys())

    def __str__(self) -> str:
        """String representation of the XBRL instance."""
        return f"XBRL instance for {self.filing.form_type} filing (CIK: {self.cik})"

    def __repr__(self) -> str:
        """Detailed string representation of the XBRL instance."""
        return (
            f"XBRLInstance(cik='{self.cik}', form='{self.filing.form_type}', "
            f"date='{self.filing.filing_date}')"
        )
